sdk/core/core/src/az_http_policy_retry.c

```diff
@@ -0,0 +1,213 @@
+            *retry_after_msec = (seconds < (INT32_MAX / _az_TIME_MILLISECONDS_PER_SECOND))
+                ? (int32_t)seconds * _az_TIME_MILLISECONDS_PER_SECOND
+                : INT32_MAX;
+
+            return AZ_OK;
+          }
+          // TODO: Other possible value is HTTP Date. For that, we'll need to parse date, get
+          // current date, subtract one from another, get seconds. And the device should have a
+          // sense of calendar clock.
+        }
+      }
+
+      *retry_after_msec = -1;
+      return AZ_OK;
+    }
+    default:
+      // None of the retriable HTTP codes above
+      *should_retry = false;
+      *retry_after_msec = -1;
+      return AZ_OK;
+  }
+}
+
+AZ_NODISCARD AZ_INLINE int32_t _az_http_policy_retry_calc_delay(
+    int16_t attempt,
+    int32_t retry_delay_msec,
+    int32_t max_retry_delay_msec) {
+  int32_t const exponential_retry_after = (int32_t)(
+      (double)retry_delay_msec
+      * (double)(attempt <= 30 ? (1 << attempt) : INT32_MAX)); // scale exponentially
+  // Multiply the above by "(((1.0 / RAND_MAX) * rand()) / 2 + 0.8)" to add jitter factor from
+  // 0.8 to 1.3
+
+  return exponential_retry_after > max_retry_delay_msec ? max_retry_delay_msec
+                                                        : exponential_retry_after;
+}
+
+AZ_NODISCARD az_result
+_az_http_policy_retry(az_http_policy_retry_options * options, az_http_policy_callback next_policy) {
+  int16_t const max_tries = options->max_tries;
+  int32_t const retry_delay_msec = options->retry_delay_msec;
+  int32_t const max_retry_delay_msec = options->max_retry_delay_msec;
+
+  az_http_pipeline_policy_func const next_http_policy = next_policy.func;
+  az_http_policy * const p_policies = next_policy.params.p_policies;
+  az_http_request_builder * const hrb = next_policy.params.hrb;
+  az_http_response * const response = next_policy.params.response;
+
+  bool const should_log = az_log_should_write(AZ_LOG_HTTP_RETRY);
+  az_result result = AZ_OK;
+  int16_t attempt = 1;
+  while (true) {
+    AZ_RETURN_IF_FAILED(az_http_response_reset(response));
+    AZ_RETURN_IF_FAILED(az_http_request_builder_remove_retry_headers(hrb));
+    result = next_http_policy(p_policies, hrb, response);
+
+    // Even HTTP 429, or 502 are expected to be AZ_OK, so the failed result is not retriable.
+    if (attempt > max_tries || az_failed(result) || response->builder.length == 0) {
+      return result;
+    }
+
+    int32_t retry_after_msec = -1;
+    bool should_retry = false;
+    AZ_RETURN_IF_FAILED(_az_http_policy_retry_get_retry_after(
+        az_span_builder_result(&response->builder), &should_retry, &retry_after_msec));
+
+    if (!should_retry) {
+      return result;
+    }
+
+    ++attempt;
+
+    if (retry_after_msec < 0) { // there wasn't any kind of "retry-after" response header
+      retry_after_msec
+          = _az_http_policy_retry_calc_delay(attempt, retry_delay_msec, max_retry_delay_msec);
```

Comment:
I can only find `_az_http_policy_retry_calc_delay` here: any reason it does the calculation as a `double` since the result is converted to an `int32_t` anyway? 
That immediately makes it extremely heavy (size and execution time) for anything that doesn't have an FPU.

Reply:
I'll rewrite this piece to not use FPU arithmetic, it looks avoidable.